src/rules/banParensRule.ts

```diff
@@ -0,0 +1,252 @@
+            return {
+                message: `the ${whichChild} child of an expression${parentKind === "*" ? "" : ` of type ${parentKind}`}`,
+                test(node: ts.ParenthesizedExpression | ts.ParenthesizedTypeNode) {
+                    if (!isNodeOfKind(node.parent!, parentKind)) {
+                        return false;
+                    }
+                    const parentMapping = node.parent as {} as { [k: string]: ts.Node | ts.Node[] };
+                    const childOrChildren = parentMapping[whichChild];
+                    return Array.isArray(childOrChildren) ?
+                        childOrChildren.indexOf(node) !== -1 :
+                        childOrChildren === node;
+                },
+            };
+        }));
+
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if ((isParenthesizedExpression(node) && !parensAreNecessary(node, ctx.sourceFile)) || isParenthesizedType(node)) {
+            const restriction = restrictions.find((r) => r.test(node));
+            if (restriction != undefined) {
+                let replacement = [
+                    Lint.Replacement.deleteFromTo(node.getStart(), node.getStart() + 1),
+                    Lint.Replacement.deleteFromTo(node.getEnd() - 1, node.getEnd()),
+                ];
+                const charBeforeParens = ctx.sourceFile.text[node.getStart() - 1];
+                // Prevent correcting typeof(x) to typeofx, throw(err) to throwerr
+                if (charBeforeParens.match(/\w/) !== null) {
+                    replacement.push(Lint.Replacement.appendText(node.getStart(), " "));
+                }
+                // Don't suggest a fix for a (hopefully rare) pattern where
+                // removing the parentheses would almost always be bad, e.g.
+                // let x = (y = 1, z = 2);
+                if (isParenthesizedExpression(node) &&
+                    isBinaryExpression(node.expression) &&
+                    node.expression.operatorToken.kind === ts.SyntaxKind.CommaToken) {
+                    replacement = [];
+                }
+                ctx.addFailureAtNode(
+                    node,
+                    Rule.FAILURE_STRING_FACTORY(restriction.message),
+                    replacement);
+            }
+        }
+        return ts.forEachChild(node, cb);
+    });
+}
+
+/**
+ * Checks some exceptional cases where the parentheses likely are still required.
+ */
+function parensAreNecessary(node: ts.ParenthesizedExpression, sourceFile: ts.SourceFile) {
+    return (
+        // Don't flag `(0).foo()`, because `0.foo()` doesn't work.
+        (isNumericLiteral(node.expression) &&
+            isPropertyAccessExpression(node.parent!)) ||
+        // Don't flag `return (\nfoo)`, since the parens are necessary.
+        (isReturnStatement(node.parent!) &&
+            !isSameLine(sourceFile, node.getStart(), node.expression.getStart())) ||
```

Comment:
please pass the `sourceFile` parameter to `.getStart()` to avoid unnecesary lookups.
I'm also very confident that you can replace `node.getStart()` with `node.expression.pos`. The former gives you the start of the open paren, the latter gives the end of the open paren. But that difference doesn't matter for `isSameLine`

Reply:
Done. Sorry, I wrestled with this a bit until I got something working and then didn't pay more attention. I didn't really know the difference between `.getStart` and `.pos` (though I think I get it now). Is there documentation for this stuff somewhere?

What's the benefit of passing `sourceFile` in?

Reply:
`.pos` is the position where the parser started to parse the current node (or token). `node.pos` is the same value as the preceding token's `.end`.
`node.getStart()` is the position where the first token of a node starts. The difference between `.pos` and `.getStart()` is caused by trivia (whitespace and comments).
Also note that `node.pos === node.getFullStart()` and `node.end === node.getEnd()`

Re: passing `sourceFile` parameter:
I wrote this down in https://palantir.github.io/tslint/develop/custom-rules/performance-tips.html
> There are serveral methods that have an optional parameter sourceFile. Don’t omit this parameter if you care for performance. If ommitted, typescript needs to walk up the node’s parent chain until it reaches the SourceFile. This can be quite costly when done frequently on deeply nested nodes.

You also find there why I prefer  not to call `.getStart()`:
> node.getStart() scans the source to skip all the leading trivia. Although barely noticeable, this operation is not for free. If you need the start position of a node more than once per function, consider caching it.

Reply:
Got it, makes sense. Thanks for the tips!